security/s2a/internal/crypter/counter.go

```diff
@@ -0,0 +1,42 @@
+package crypter
+
+import "errors"
+
+const counterLen = 12
+
+// counter is a 96-bit, little-endian counter.
+type counter struct {
+	value   [counterLen]byte
+	invalid bool
+}
+
+func newCounter() counter {
+	return counter{}
+}
+
+// val returns the current value of the counter as a byte slice.
```

Comment:
I think the `mask()` function should be part of the S2A Half Connection rather than the Counter. The Counter right now provides a method to easily retrieve a byte slice representation of the sequence and also provides overflow checking. I think exposing a `mask` method would break the abstraction that the counter represents so I think it belong to S2A Half Connection rather than the Counter. WDYT?

Reply:
SGTM.